### PR TITLE
refactor: Wallet derivation

### LINKED ISSUE #87: refactor: Add wallet derivation in `transaction_signing_gateway`

### PR DESCRIPTION
**Pull request by wal33d006**
Fixes #87

### CODE REVIEW

File: packages/transaction_signing_gateway/lib/wallet_derivator.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import 'package:transaction_signing_gateway/model/network_info.dart';
+import 'package:transaction_signing_gateway/model/private_wallet_credentials.dart';
+import 'package:transaction_signing_gateway/model/wallet_derivation_failure.dart';
+
+abstract class WalletDerivator {
+  Either<WalletDerivationFailure, PrivateWalletCredentials> derive({
+    required String mnemonic,
+    required String walletAlias,
+    required NetworkInfo networkInfo,
+  });
```
Comment (andrzejchm) [inline packages/transaction_signing_gateway/lib/wallet_derivator.dart:11]: we cannot accept NetworkInfo here, as this model is specific to Alan and Cosmos network. what we can do instead is to accept following:

```suggestion
  Either<WalletDerivationFailure, PrivateWalletCredentials> derive({
    required WalletDerivationInfo walletDerivationInfo
  });
```
Add new interface for `WalletDerivationInfo`:

```dart
abstract class WalletDerivationInfo {
}
```

and implementation specific for Alan

```dart
class AlanWalletDerivationInfo implements WalletDerivationInfo {
    final String mnemonic;
    final String walletAlias;
    final NetworkInfo networkInfo; // Alan's version of network info, we don't need a separate model in the transaction_signing_gateway library
    
}
```

File: packages/transaction_signing_gateway/lib/model/network_info.dart [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import 'package:equatable/equatable.dart';
+
+class NetworkInfo extends Equatable {
```
Comment (andrzejchm) [inline packages/transaction_signing_gateway/lib/model/network_info.dart:3]: we can remove that after suggested changes are in place

File: packages/transaction_signing_gateway/lib/wallet_derivator.dart [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import 'package:dartz/dartz.dart';
+import 'package:transaction_signing_gateway/model/network_info.dart';
+import 'package:transaction_signing_gateway/model/private_wallet_credentials.dart';
+import 'package:transaction_signing_gateway/model/wallet_derivation_failure.dart';
+
+abstract class WalletDerivator {
```
Comment (andrzejchm) [inline packages/transaction_signing_gateway/lib/wallet_derivator.dart:6]: We would also need to add a `derive` method inside the `TransactionSigningGateway` class and make it work similar to what we have for transaction signing, where we have a list of derivators and look for one which is capable of handling the derivation

**Review by andrzejchm — CHANGES_REQUESTED** [review]

File: packages/transaction_signing_gateway/lib/wallet_derivator.dart (hunk 3 above)
Comment (wal33d006) [reply packages/transaction_signing_gateway/lib/wallet_derivator.dart:6]: ✅

File: packages/transaction_signing_gateway/lib/model/network_info.dart (hunk 2 above)
Comment (wal33d006) [reply packages/transaction_signing_gateway/lib/model/network_info.dart]: ✅

File: packages/transaction_signing_gateway/lib/wallet_derivator.dart (hunk 1 above)
Comment (wal33d006) [reply packages/transaction_signing_gateway/lib/wallet_derivator.dart]: ✅

**Review by andrzejchm — CHANGES_REQUESTED** [review]

File: packages/transaction_signing_gateway/lib/gateway/transaction_signing_gateway.dart [hunk 4]
```diff
@@ -78,6 +85,10 @@ class TransactionSigningGateway {
                 privateWalletCredentials: privateCreds,
               ));
 
+  Future<Either<WalletDerivationFailure, PrivateWalletCredentials>> deriveWallet(
+          {required AlanWalletDerivationInfo alanWalletDerivationInfo}) async =>
```
[the remark this answers is not in the document]
Comment (wal33d006) [reply packages/transaction_signing_gateway/lib/gateway/transaction_signing_gateway.dart]: ✅

**Review by andrzejchm — APPROVED** [review]
🚀

**Pull request merged**